File: python/src/tools/insights/update.py
import json
from pydantic import BaseModel
from src.tools.types import Context, Tool, ToolResult, TextContent
from src.schema.insights import UpdateInsightInput
from src.lib.utils.api import get_project_base_url


class UpdateInsightParams(BaseModel):
    insightId: int
    data: UpdateInsightInput


async def update_insight_handler(context: Context, params: UpdateInsightParams) -> ToolResult:
    project_id = await context.get_project_id()
    insight_result = await context.api.insights(project_id).update(params.insightId, params.data)

    if not insight_result.success:
        raise Exception(f"Failed to update insight: {insight_result.error}")

    insight_with_url = {
        **insight_result.data.model_dump(),
        "url": f"{get_project_base_url(project_id)}/insights/{insight_result.data.short_id}"
    }

    return ToolResult(content=[TextContent(text=json.dumps(insight_with_url))])


def update_insight_tool() -> Tool[UpdateInsightParams]:
    return Tool(
        name="insight-update",
        description="""
        - Update an existing insight by ID.
        - Can update name, description, filters, and other properties.
        """,
        schema=UpdateInsightParams,
        handler=update_insight_handler
    )

File: typescript/src/tools/surveys/update.ts
import { SurveyUpdateSchema } from "@/schema/tool-inputs";
import { formatSurvey } from "@/tools/surveys/utils/survey-utils";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = SurveyUpdateSchema;
type Params = z.infer<typeof schema>;

export const updateHandler = async (context: Context, params: Params) => {
	const { surveyId, ...data } = params;
	const projectId = await context.stateManager.getProjectId();

	if (data.questions) {
		data.questions = data.questions.map((question: any) => {
			// Handle single choice questions - convert numeric keys to strings
			if (
				"branching" in question &&
				question.branching?.type === "response_based" &&
				question.type === "single_choice"
			) {
				question.branching.responseValues = Object.fromEntries(
					Object.entries(question.branching.responseValues).map(([key, value]) => {
						return [String(key), value];
					}),
				);
			}
			return question;
		});
	}

	const surveyResult = await context.api.surveys({ projectId }).update({
		surveyId,
		data,
	});

	if (!surveyResult.success) {
		throw new Error(`Failed to update survey: ${surveyResult.error.message}`);
	}

	const formattedSurvey = formatSurvey(surveyResult.data, context, projectId);

	return {
		content: [{ type: "text", text: JSON.stringify(formattedSurvey) }],
	};
};

const definition = getToolDefinition("survey-update");

const tool = (): Tool<typeof schema> => ({
	name: "survey-update",
	description: definition.description,
	schema,
	handler: updateHandler,
	annotations: {
		destructiveHint: false,
		idempotentHint: false,
		openWorldHint: true,
		readOnlyHint: false,
	},
});

export default tool;
